### ISSUE TITLE
Spotify playback does not start after long idle period on a Google Cast device

### ISSUE THREAD

**Issue by bithajcsar** [fondberg/spotcast#298]
I am experiencing a very strange behavior:
My automation containing Spotify playback on a Google Cast device works well always except after long (6-8 hours) idle period. The automation itself is a wake-up flow triggered by a time reached. The playback should start at the specified time. And it works always except in the morning when the only active device is the HA server (RasPi) and all other devices having Spotify clients are off for a longer time. The problem is exceptionally annoying as far as everything is working except the wake-up when we really need that.
At the wake-up time the Spotify client on the Google Cast device dos not start (black screen) and I see the following in the log:

```log
File "/usr/local/lib/python3.9/site-packages/spotipy/client.py", line 267, in _internal_call
    raise SpotifyException(
spotipy.exceptions.SpotifyException: http status: 403, code:-1 - https://api.spotify.com/v1/me/player/play?device_id=xxxxxxxxxxxx:
 Player command failed: Restriction violated, reason: UNKNOWN
```

* [x] Using latest version of spotcast
* [x] I have setup the Spotify integration in Home Assistant
* [ ] I am using multiple accounts
* [x] Using latest stable version of Home Assistant
* [x] I have Spotify Premium
* [ ] I'm attaching relevant logs with level debug for component spotcast (see README)
* [ ] I'm using entity_id in the service call and have tried device_name but the issue remains

**Environment (please complete the following information):**
 - Installation type: HA_OS
 - HA version: core-2022.2.5
 - spotcast version: 3.6.24

**Configuration**

```yaml
spotcast:
  sp_dc: !secret sp_dc
  sp_key: !secret sp_key
```

**Additional context**

**Comment by fcusson** [comment]
@bithajcsar, The error is raised in Spotipy, not Spotcast, we would need a debug log to verify what is going on and if the problem originates from this project or the dependency. Also, could you provide the automation you are using? We could see what is going on there.

**Comment by bithajcsar** [comment]
I was not able to reproduce the issue yet.

**Comment by odtgit** [comment]
I have a spotcast call within my morning routine in NR that has failed like this two mornings in row with the same 403 error according to the log. Any suggestions what I can do to work around it? I have a separate media_player.set_volume call before so the speaker is up and running.
```json

{
    "entity_id": "media_player.bedroom_speaker",
    "uri": "spotify:playlist:37i9dQZEVXcHpyiKpvAF8k",
    "random_song": "true",
    "shuffle": "true"
}
```
```

2022-02-16 08:00:00 INFO (SyncWorker_1) [pychromecast.controllers] Receiver:setting volume to 0.1
2022-02-16 08:00:01 INFO (SyncWorker_3) [pychromecast.controllers] Receiver:Launching app CC32E753
2022-02-16 08:00:05 ERROR (SyncWorker_3) [spotipy.client] HTTP Error for PUT to https://api.spotify.com/v1/me/player/play?device_id=e94cf90cb9d45f8f433521899bc85928 with Params: {} returned 403 due to Player command failed: Restriction violated
2022-02-16 08:00:05 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [140676828626176] http status: 403, code:-1 - https://api.spotify.com/v1/me/player/play?device_id=e94cf90cb9d45f8f433521899bc85928:
 Player command failed: Restriction violated, reason: UNKNOWN
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/spotipy/client.py", line 245, in _internal_call
    response.raise_for_status()
  File "/usr/local/lib/python3.9/site-packages/requests/models.py", line 960, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 403 Client Error: Forbidden for url: https://api.spotify.com/v1/me/player/play?device_id=e94cf90cb9d45f8f433521899bc85928

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 190, in handle_call_service
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 1630, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1673, in _execute_service
    await self._hass.async_add_executor_job(
  File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/spotcast/__init__.py", line 215, in start_casting
    spotcast_controller.play(
  File "/config/custom_components/spotcast/spotcast_controller.py", line 275, in play
    client.start_playback(**kwargs)
  File "/usr/local/lib/python3.9/site-packages/spotipy/client.py", line 1767, in start_playback
    return self._put(
  File "/usr/local/lib/python3.9/site-packages/spotipy/client.py", line 312, in _put
    return self._internal_call("PUT", url, payload, kwargs)
  File "/usr/local/lib/python3.9/site-packages/spotipy/client.py", line 267, in _internal_call
    raise SpotifyException(
spotipy.exceptions.SpotifyException: http status: 403, code:-1 - https://api.spotify.com/v1/me/player/play?device_id=e94cf90cb9d45f8f433521899bc85928:
 Player command failed: Restriction violated, reason: UNKNOWN

```

**Comment by hmmbob** [comment]
Could this be caused by https://github.com/home-assistant/core/issues/66373?

**Comment by odtgit** [comment]
This only happens in the morning though, throughout the day spocast calls are working flawlessly, My guess something with an orphaned device?

**Comment by fcusson** [comment]
@odtgit 

so the problem is with the device access url. Spotify does understand your request, knows the device exist, but believe you do not have the right to access the device itself.

> 403 Forbidden
The HTTP 403 Forbidden response status code indicates that the server understands the request but refuses to authorize it.

> This status is similar to [401](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/401), but for the 403 Forbidden status code re-authenticating makes no difference. The access is permanently forbidden and tied to the application logic, such as insufficient rights to a resource.

Unless Spotify makes bad uses of error codes it shouldn't be a token issue, because a token issue should spit back a `401` which is for bad/deprecated/expired authentication

Could you provide more information on the device that is used (manufacturer, ecosystem, etc?) Also we could try to call the service with the device ID directly, there might be an issue in the method matching the spotify ID. Could you follow [this procedure](https://github.com/fondberg/spotcast#find-spotify-device-id) and check those elements:

- that you are able to start playback using the Spotify ID
- if the Spotify ID matches with the one in the uri `e94cf90cb9d45f8f433521899bc85928`

**Comment by odtgit** [comment]
It's a JBL link portable using the google cast protocol. It's always listed in my desktop spotify client as a target but *not* available after a while on the developer console output/API.

Using a spotify_device_id in spotcast.start when it's been off returns a 404 quickly as expected since it's not in the listed players. When using the entity_id like I have above it will start playing after a couple seconds. 

```
{
  "devices": [
    {
      "id": "e94cf90cb9d45f8f433521899bc85928",
      "is_active": true,
      "is_private_session": false,
      "is_restricted": false,
      "name": "JBL",
      "type": "CastAudio",
      "volume_percent": 40
    },
```

**Comment by fcusson** [comment]
ok so very strange, the ID is the same but still its forbidden from Spotify API.

The fact it goes *not available* after a while might be linked to your problem. It is a real device so it doesn't send `404` since the link does exist. It doesn't send `401` because your token is valid. So it probably send `403` since it is unable to communicates with the device.

Since its server is working correctly it doesn't go into the `5XX` error space. Please try with a media_player.turn_on. If it does work, we might have a way to implement a similar work around in the code that we could use when an HTTPError is raised

**Comment by odtgit** [comment]
Thanks for the explanation. I think using the entity_id playback is already doing some magic behind the scenes as this flow has been working fine for a long time now with just a media_player.volume_set before spotcast.start but that isn't the case as of this week any longer. I've upgraded to the latest spotcast (3.6.25) and added the turn_on call. I'll report back if it works fine the rest of the week.

**Comment by odtgit** [comment]
Hi,

The issue is fixed after including media_plsyer.turn_on before the
spotcast.start call.

**Comment by bithajcsar** [comment]
Although I also added media player turn on call before calling the spotcast start, the playback did not start in the morning today. Caused some problem...
Unfortunately I could not see log entries on the HA GUI due to another integration producing a lot of log entries.

**Comment by MattDemers** [comment]
Had this issue this morning as well.

**Comment by fcusson** [comment]
@MattDemers would you have any luck in providing logs or details (device used, the script, etc)

**Comment by MattDemers** [comment]
> @MattDemers would you have any luck in providing logs or details (device used, the script, etc)

Will do when I get home. I found this thread by searching for a 403 errors; I got the same error as this ```
http status: 403, code:-1 - https://api.spotify.com/v1/me/player/play?device_id=xxxxxxxxxxxx:
 Player command failed: Restriction violated, reason: UNKNOWN
```

At least the bits about the Restriction violated.

**Comment by MattDemers** [comment]
```
Stopped because an error was encountered at March 9, 2022, 8:43:03 AM (runtime: 3.83 seconds)

http status: 403, code:-1 - https://api.spotify.com/v1/me/player/play?device_id=****** (obfuscated by me): Player command failed: Restriction violated, reason: UNKNOWN
```

Used "Call Service" in my script, with a playlist URI, "Living Room Speaker" being my my Google Chromecast Audio, Random Song, Shuffle both being on, and a Start Volume of 75.

Clicking "Run Script" now, it works fine. Will see if it errors tomorrow, too.

Since I'm using it as a morning alarm, I'd obviously prefer this to be reliable.

**Comment by fcusson** [comment]
Should be resolved in `v3.7.0`. Closing please reopne if still existing

**Comment by kingp0dd** [comment]
I'm experiencing the same issue even on v3.7.0

This only happens when the device is waking up the first time from a deep sleep.

I tried media_player.turn_on and media_player.play_media (HA Spotify integration) before the spotcast command but I still get the following error.

```
Log Details (ERROR)
This error originated from a custom integration.
Logger: homeassistant.core
Source: custom_components/spotcast/spotify_controller.py:104
Integration: Spotcast (documentation, issues)
First occurred: June 9, 2023 at 8:02:07 AM (19 occurrences)
Last logged: 8:06:14 AM

Error executing service: <ServiceCall script.spotify_kygo_piano (c:01H3DKT2ER27SA9Y104DBJFYRJ)>
Error executing service: <ServiceCall spotcast.start (c:01H3DKT4EZ548DH90TKZRVMW3F): entity_id=media_player.speakers, uri=spotify:playlist:5jyOi542k447ls4MHoHdf79R, shuffle=False, force_playback=False, start_volume=101, ignore_fully_played=False, repeat=off, offset=0, random_song=False, limit=20>
Error executing service: <ServiceCall spotcast.start (c:01H3DM89ANFVFGND9ZG4EXPHNN): entity_id=media_player.speakers, uri=spotify:playlist:5jyOi5424k57ls4MHoHdf79R, shuffle=False, force_playback=False, start_volume=101, ignore_fully_played=False, repeat=off, offset=0, random_song=False, limit=20>
Error executing service: <ServiceCall spotcast.start (c:01H3DMARY9WV6CBRXVA4SN6YVB): entity_id=media_player.speakers, uri=spotify:playlist:5jyOi542ne547ls4MHoHdf79R, shuffle=False, force_playback=False, start_volume=101, ignore_fully_played=False, repeat=off, offset=0, random_song=False, limit=20>
Error executing service: <ServiceCall spotcast.start (c:01H3DMBVNNEB7MGZWAHM58CM6B): entity_id=media_player.speakers, uri=spotify:playlist:5jyOi54mr247ls4MHoHdfjs79R, shuffle=False, force_playback=False, start_volume=101, ignore_fully_played=False, repeat=off, offset=0, random_song=False, limit=20>
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/core.py", line 1928, in catch_exceptions
    await coro_or_task
  File "/usr/src/homeassistant/homeassistant/core.py", line 1956, in _execute_service
    await self._hass.async_add_executor_job(
  File "/usr/local/lib/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/spotcast/__init__.py", line 221, in start_casting
    spotify_device_id = spotcast_controller.get_spotify_device_id(
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/spotcast/spotcast_controller.py", line 250, in get_spotify_device_id
    spotify_cast_device.startSpotifyController(access_token, expires)
  File "/config/custom_components/spotcast/spotcast_controller.py", line 95, in startSpotifyController
    sp.launch_app()
  File "/config/custom_components/spotcast/spotify_controller.py", line 104, in launch_app
    raise LaunchError(
pychromecast.error.LaunchError: Timeout when waiting for status response from Spotify app

```
